Add Control4 integration

lawtancool opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds a Control4 integration that supports lights only. 

[Control4](https://www.control4.com/) is a home automation system that uses a local controller/hub to interface with lights, thermostats, alarm systems, etc.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13941

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

lawtancool committed:
initial commit

lawtancool committed:
working init and config flow

lawtancool committed:
config flow that adds device

lawtancool committed:
working light

lawtancool committed:
fix entity and device registry
needed to add unique_id

lawtancool committed:
fix translations

lawtancool committed:
update hass requirements

lawtancool committed:
add control4 to ssdp

lawtancool committed:
omit control4 in coveragerc

lawtancool committed:
black formatting - remove trailing newline

lawtancool committed:
add docstrings and remove useless imports

lawtancool committed:
fix translation json comma

lawtancool committed:
add light transition time options

lawtancool committed:
cleanup unused imports and docstrings

lawtancool committed:
order imports with isort

lawtancool committed:
raise PlatformNotReady if cannot connect to Control4 API

homeassistant commented:
Hi @lawtancool, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#37632).

Once you do that we will be able to review and accept this pull request.

Thanks!

lawtancool committed:
fix typo in .coveragerc

lawtancool committed:
fix import sorting in init

lawtancool committed:
remove unused imports in init

lawtancool committed:
fix import sorting in init

lawtancool committed:
add docstrings to satisfy flake8

lawtancool committed:
fix config_flow test

lawtancool committed:
update coveragerc for added config_flow test

lawtancool committed:
remove reliance on entry.title
switch to referencing entry.entry_id,  store controller_name seperately

lawtancool committed:
add more variable names to const.py

lawtancool committed:
update config_flow test
addresses adding controller_name to data

lawtancool committed:
remove placeholder docstring in config_flow

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,275 @@
+    """Set up Control4 from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN].setdefault(entry.entry_id, {})
```
```suggestion
    entry_data = hass.data[DOMAIN].setdefault(entry.entry_id, {})
```

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,275 @@
+    except client_exceptions.ClientError as exception:
+        _LOGGER.error("Error connecting to Control4 account API: %s", exception)
+        raise PlatformNotReady()
+    hass.data[DOMAIN][entry.entry_id][CONF_ACCOUNT] = account
```
```suggestion
    entry_data[CONF_ACCOUNT] = account
```

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,275 @@
+    hass.data[DOMAIN][entry.entry_id][CONF_ACCOUNT] = account
+
+    controller_name = config[CONF_CONTROLLER_NAME]
+    hass.data[DOMAIN][entry.entry_id][CONF_CONTROLLER_NAME] = controller_name
```
```suggestion
    entry_data[CONF_CONTROLLER_NAME] = controller_name
```

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,275 @@
+
+    director_token_dict = await account.getDirectorBearerToken(controller_name)
+    director = C4Director(config[CONF_HOST], director_token_dict[CONF_TOKEN])
+    hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR] = director
```
```suggestion
    entry_data[CONF_DIRECTOR] = director
```

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,275 @@
+    director_token_dict = await account.getDirectorBearerToken(controller_name)
+    director = C4Director(config[CONF_HOST], director_token_dict[CONF_TOKEN])
+    hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR] = director
+    hass.data[DOMAIN][entry.entry_id][
```
```suggestion
    entry_data[
```

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,275 @@
+
+    # Add Control4 controller to device registry
+    controller_href = (await account.getAccountControllers())["href"]
+    hass.data[DOMAIN][entry.entry_id][
```
```suggestion
    entry_data[
```

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,275 @@
+    ] = await account.getControllerOSVersion(controller_href)
+
+    result = re.search("_(.*)_", controller_name)
+    hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR_MODEL] = result.group(1).upper()
```
```suggestion
    entry_data[CONF_DIRECTOR_MODEL] = result.group(1).upper()
```

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,275 @@
+
+    Configuration through YAML is not supported at this time.
+    """
+    return True
```
```suggestion
    hass.data.setdefault(DOMAIN, {})
    return True
```

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,275 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Control4 from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
```
```suggestion
```

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,275 @@
+    device_registry = await dr.async_get_registry(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        identifiers={(DOMAIN, controller_name)},
```
The identifier needs to be stable and something the user cannot change such as a serial number of mac address.  It should conform to the unique id requirements.   

https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements

lawtancool replied:
`controller_name` is assigned by the Control4 API based on MAC address - as far as I know, it cannot be changed by the user.

lawtancool committed:
Apply suggestions from code review

Replace hass.data[DOMAIN][entry.entry_id] with entry_data = hass.data[DOMAIN].setdefault(entry.entry_id, {})

Co-authored-by: J. Nick Koston <nick@koston.org>

lawtancool committed:
More code review changes

lawtancool commented:
Thanks for the review @bdraco! I've applied your suggestions.

lawtancool commented:
Moved to draft because I discovered a bug where tokens were not being refreshed properly.

lawtancool committed:
Attempt to fix token updating
move update logic to update coordinator instead of async_update

lawtancool committed:
bump requirements to pyControl4==0.0.4

lawtancool committed:
flake8 fixes and added docstrings

lawtancool committed:
add docstring to director_utils.py

lawtancool committed:
Add note to director_update_data

lawtancool committed:
refresh_tokens account bearer token fix

lawtancool committed:
remove todo to pass flake8

bdraco commented on homeassistant/components/control4/light.py:
```diff
@@ -0,0 +1,223 @@
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities
+):
+    """Set up Control4 lights from a config entry."""
+    scan_interval = hass.data[DOMAIN][entry.entry_id][CONF_SCAN_INTERVAL]
```
```suggestion
    entry_data = hass.data[DOMAIN][entry.entry_id]
    scan_interval = entry_data[CONF_SCAN_INTERVAL]
```

....

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,233 @@
+    ):
+        """Initialize a Control4 entity."""
+        self.entry = entry
+        self.account = hass.data[DOMAIN][self.entry.entry_id][CONF_ACCOUNT]
```
```suggestion
        entry_data = hass.data[DOMAIN][self.entry.entry_id]
        self.account = entry_data[CONF_ACCOUNT]
```

....

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,205 @@
+            self.director = C4Director(self.host, self.director_bearer_token)
+            await self.director.getAllItemInfo()
+            return True
+        except Exception as exception:  # pylint: disable=broad-except
```
Is there another way to tell we can't connect besides a broad exception trap?

lawtancool replied:
I've changed this to catch `aiohttp` Client exceptions and the generic `asyncio` TimeoutError, as I found those were the errors I encountered when I entered an incorrect IP address (that wasn't a Control4 controller). 

Is that acceptable?

bdraco replied:
Works for me.

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,205 @@
+                self.controller_name
+            )
+            return True
+        except (BadCredentials, Unauthorized) as exception:
+            _LOGGER.error(exception)
```
This is expected to happen.  We shouldn't log the exception especially if it contains credentials

```suggestion
        except (BadCredentials, Unauthorized):
```

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,205 @@
+    if not await hub.connect_to_director():
+        raise CannotConnect
+
+    # If you cannot connect:
+    # throw CannotConnect
+    # If the authentication is wrong:
+    # InvalidAuth
```
Stale comments

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,205 @@
+            try:
+                info = await validate_input(self.hass, user_input)
+                if info["controller_name"] in configured_instances(self.hass):
+                    return self.async_abort(reason="already_configured")
```
Abort inside the try handler will not abort because it will be caught by the broad exception

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,205 @@
+        """Handle the initial step."""
+        errors = {}
+        if user_input is not None:
+
```
Does the controller have a unique id?

```
await self.async_set_unique_id(info.serial_number)
self._abort_if_unique_id_configured()
```

https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements

lawtancool replied:
Yep, `controller_name` is actually unique and cannot be changed by the user (I'm pretty sure - Control4 only allows certified dealers to access controller config, so I can't test this)

bdraco replied:
If the dealer can change it, it is not exactly stable.  Is there something better we can use?

lawtancool replied:
Unfortunately there isn't - this is the way the Control4 API/apps uniquely identify controllers, so I think that's the best way to do it.

bdraco replied:
We can't use something that we configure as it doesn't meet the unique id requirements.

Its better to not set a unique id in this case and check to see if an existing entry has the same controller name instead.

`for entry in self._async_current_entries()`

bdraco replied:
It looks like the mac address is provided here ?
https://github.com/lawtancool/pyControl4/blob/master/pyControl4/account.py#L124

lawtancool replied:
Yep, that `"controllerCommonName": "control4_MODEL_MACADDRESS",` string is what I'm using as the unique id. I'm pretty sure this is unique, it's just that the Control4 API is not publicly documented so I can't confirm.

bdraco replied:
What does that string actually look like in production ?

lawtancool replied:
It looks something like `control4_hc800_000FFF500000`

bdraco replied:
Is the mac address the actual mac address of the device on your network ?

lawtancool replied:
Yes it is.

bdraco replied:
Perfect,

```

from homeassistant.helpers.device_registry import format_mac


control4, model, mac = controller_common_name.split('_', 3)
formatted_mac = format_mac(mac)
```


Then use the `formatted_mac` for the uniqueid

bdraco commented on homeassistant/components/control4/director_utils.py:
```diff
@@ -0,0 +1,58 @@
+    director_token_expiry = director_token_dict["token_expiration"]
+
+    _LOGGER.debug("Saving new tokens in config_entry")
+    hass.data[DOMAIN][entry.entry_id][CONF_ACCOUNT] = account
```
Please write this as `entry_data = ....` any time you are using `hass.data[DOMAIN][entry.entry_id]` multiple times

lawtancool committed:
code quality improvements
properly handle already_configured config_flow case

lawtancool committed:
fix config_flow error handling

lawtancool committed:
remove unused import in config_flow

lawtancool committed:
catch TimeoutError as cannot_connect

lawtancool committed:
fix config_flow import sorting

lawtancool committed:
avoid redefining built-in TimeoutError

lawtancool committed:
try to fix asyncio import to pass tests

lawtancool committed:
remove import from asyncio.exceptions
apparently exceptions are supposed to be imported directly from asyncio

lawtancool committed:
fix config_flow tests

lawtancool committed:
config_flow docstring format fix

lawtancool committed:
code review fixes

lawtancool commented:
I've resolved the token refresh issue, and the code works correctly on my local system.

Thanks again for the review @bdraco!

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,198 @@
+        self.host = host
+        self.username = username
+        self.password = password
+        self.account = C4Account(self.username, self.password)
```
Please move side effects out of init.  Can you create this in `authenticate` ?
```suggestion
        self.account = None
```

lawtancool committed:
remove account side effect from config_flow init

bdraco commented on tests/components/control4/test_config_flow.py:
```diff
@@ -0,0 +1,100 @@
+    assert result["type"] == "form"
+    assert result["errors"] == {}
+
+    with patch(
+        "homeassistant.components.control4.config_flow.Control4Validator.authenticate",
+        return_value=True,
+    ), patch(
+        "homeassistant.components.control4.config_flow.Control4Validator.connect_to_director",
+        return_value=True,
+    ), patch(
+        "homeassistant.components.control4.config_flow.Control4Validator.return_controller_name",
+        return_value="control4_model_00AA00AA00AA",
+    ), patch(
```
Please patch the code from pypi package instead.

Ok to patch `async_setup` and `async_setup_entry`

bdraco commented on tests/components/control4/test_config_flow.py:
```diff
@@ -0,0 +1,100 @@
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.control4.config_flow.Control4Validator.authenticate",
+        side_effect=InvalidAuth,
+    ):
```
Please patch the code from pypi package instead.

Ok to patch `async_setup` and `async_setup_entry`

bdraco commented on tests/components/control4/test_config_flow.py:
```diff
@@ -0,0 +1,100 @@
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "homeassistant.components.control4.config_flow.Control4Validator.authenticate",
+        return_value=True,
+    ), patch(
+        "homeassistant.components.control4.config_flow.Control4Validator.connect_to_director",
+        side_effect=CannotConnect,
+    ):
```
Please patch the code from pypi package instead.

Ok to patch `async_setup` and `async_setup_entry`

bdraco commented on homeassistant/components/control4/light.py:
```diff
@@ -0,0 +1,220 @@
+        self._is_dimmer = is_dimmer
+        self._transition_time = light_transition_time
+        self._cold_start_transition_time = light_cold_start_transition_time
+        # pylint: disable=invalid-name
+        self._C4Light = C4Light(self.director, idx)
```
Since this is in the core codebase and in the scope of what we can change, please use a valid name instead of disabling pylint.

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,199 @@
+            _LOGGER.error(exception)
+            return False
+
+    async def return_controller_name(self) -> str:
```
This shouldn't be a co-routine as nothing is being awaited
```suggestion
    def return_controller_name(self) -> str:
```

bdraco reviewed (changes requested):
Please see comments above.  I'll try to get another review pass in tomorrow if you are trying to get this in before beta cut Wednesday

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,230 @@
+        controller_href
+    )
+
+    result = re.search("_(.*)_", controller_name)
+    entry_data[CONF_DIRECTOR_MODEL] = result.group(1).upper()
```
```suggestion
    entry_data[CONF_DIRECTOR_MODEL] = controller_name.split('_')[1].upper()
```

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,230 @@
+    device_registry = await dr.async_get_registry(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        identifiers={(DOMAIN, controller_name)},
```
`from homeassistant.helpers import device_registry as dr`

```suggestion
        connections={(dr.CONNECTION_NETWORK_MAC, mac_address)},
```

lawtancool replied:
I'm going to retain the `identifiers` part because that identifier is used to link entities to the controller in device registry with `via_device` https://github.com/home-assistant/core/pull/37632/files#diff-d12df26ffd8ede1b79f9660721ec84faR209

bdraco replied:
Let's change `controller_name` to `controller_unique_id` so other reviewers don't go down the path of investigating this in the future as it will save a lot of time later.

lawtancool replied:
Sure, sounds good!

lawtancool committed:
don't make return_controller_name() a coroutine

lawtancool committed:
fix naming style of C4Light

lawtancool committed:
add usage of controller MAC address

lawtancool committed:
change all occurences of controller_name to controller_unique_id

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,193 @@
+            await self.async_set_unique_id(formatted_mac)
+            self._abort_if_unique_id_configured()
+
+            if errors == {}:
```
```suggestion
            if not errors:
```

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,193 @@
+            self.director = C4Director(self.host, self.director_bearer_token)
+            await self.director.getAllItemInfo()
+            return True
+        except (Unauthorized, ClientError, asyncioTimeoutError) as exception:
+            _LOGGER.error(exception)
```
```suggestion
        except (Unauthorized, ClientError, asyncioTimeoutError):
            _LOGGER.exception("Failed to connect to the director")
```

lawtancool replied:
I think it would be better to include the exception message instead - at least for the `Unauthorized` exception from the PyPI package, the exception message contains the response data from the controller, which could be useful for troubleshooting.

bdraco replied:
_LOGGER.exception will look the exception.

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,233 @@
+        await account.getAccountBearerToken()
+    except client_exceptions.ClientError as exception:
+        _LOGGER.error("Error connecting to Control4 account API: %s", exception)
+        raise PlatformNotReady()
```
```suggestion
        raise ConfigEntryNotReady
```

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,233 @@
+async def update_listener(hass, config_entry):
+    """Update when config_entry options update."""
+    _LOGGER.debug("Config entry was updated, rerunning setup")
+    await async_unload_entry(hass, config_entry)
+    await async_setup_entry(hass, config_entry)
```
```suggestion
    await hass.config_entries.async_reload(config_entry.entry_id)
```

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,233 @@
+    if unload_ok:
+        controller_name = entry.entry_id
+        hass.data[DOMAIN].pop(entry.entry_id)
+        _LOGGER.debug("Unloaded entry for %s", controller_name)
```
This isn't the name, its the config entry id

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,233 @@
+        self._coordinator = coordinator
+        self._controller_name = hass.data[DOMAIN][self.entry.entry_id][
+            CONF_CONTROLLER_NAME
+        ]
```
```suggestion
        entry_data = hass.data[DOMAIN][self.entry.entry_id]
        self.account = entry_data[CONF_ACCOUNT]
        self.director = entry_data[CONF_DIRECTOR]
        self.director_token_expiry = entry_data[CONF_DIRECTOR_TOKEN_EXPIRATION]
        self._name = name
        self._idx = idx
        self._coordinator = coordinator
        self._controller_name = entry_data[CONF_CONTROLLER_NAME]
```

lawtancool committed:
add return values of pyControl4 to test

lawtancool commented on tests/components/control4/test_config_flow.py:
```diff
@@ -0,0 +1,105 @@
+    assert result["errors"] == {}
+
+    with patch(
+        "pyControl4.account.C4Account.getAccountControllers",
```
@bdraco I've added the return values expected from the pyControl4 package, but it doesn't seem to actually work (the variables that are assigned these values are returning as `None` when I run the test)

Could you take a look and suggest how this is supposed to be done?

bdraco replied:
You need to mock the `pyControl4.account.C4Account` objection creation with a pytest fixture or something like `harmony/test_config_flow.py` does

lawtancool replied:
After looking at `harmony/test_config_flow.py`, I've gotten `test_form` to work. 
However, I can't get `test_form_invalid_auth` and `test_form_cannot_connect` to pass. Even though I'm raising the same `Unauthorized` exceptions as a `side_effect`, they don't seem to be caught properly by `config_flow`.

bdraco replied:
This should fix it https://github.com/home-assistant/core/pull/37632#discussion_r454672497

lawtancool committed:
replace entry unload/setup with reload function

Co-authored-by: J. Nick Koston <nick@koston.org>

lawtancool committed:
change =={} to not

Co-authored-by: J. Nick Koston <nick@koston.org>

lawtancool committed:
raise ConfigEntryNotReady instead of PlatformNotReady

lawtancool committed:
correctly mock C4Account and C4Director objects

bdraco commented on homeassistant/components/control4/director_utils.py:
```diff
@@ -0,0 +1,57 @@
+    director = C4Director(config[CONF_HOST], director_token_dict[CONF_TOKEN])
+    director_token_expiry = director_token_dict["token_expiration"]
+
+    _LOGGER.debug("Saving new tokens in config_entry")
```
```suggestion
    _LOGGER.debug("Saving new tokens in hass data")
```

This comment was a bit confusing since don't actually write the tokens in the config entry to `.storage/core.config_entries`

lawtancool committed:
correct debug statement when saving tokens

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented on homeassistant/components/control4/light.py:
```diff
@@ -0,0 +1,219 @@
+
+    items_of_category = await get_items_of_category(hass, entry, CONTROL4_CATEGORY)
+    for item in items_of_category:
+        if item["type"] == 7:
```
What does 7 mean here?  Can we make it a constant so we show what the magic number represents?
```suggestion
        if item["type"] == MAYBE_SOME_CONSTANT:
```

lawtancool replied:
`7` is a Control4 item type that's analogous to Home Assistant entities (one device [type `6`] can have multiple entities [type `7`]). I'll probably add this to `const.py`

bdraco commented on homeassistant/components/control4/light.py:
```diff
@@ -0,0 +1,219 @@
+        self._is_dimmer = is_dimmer
+        self._transition_time = light_transition_time
+        self._cold_start_transition_time = light_cold_start_transition_time
+        self._c4_light = C4Light(self.director, idx)
```
Please move side effects out of `__init__`

Myabe create this in `async def async_added_to_hass`

lawtancool committed:
add CONTROL4_ENTITY_TYPE constant

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,193 @@
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+            controller_unique_id = hub.return_controller_unique_id()
+            control4, model, mac = controller_unique_id.split("_", 3)
+            formatted_mac = format_mac(mac)
+            await self.async_set_unique_id(formatted_mac)
+            self._abort_if_unique_id_configured()
+
+            if not errors:
+                return self.async_create_entry(
+                    title=controller_unique_id,
+                    data={
+                        CONF_HOST: user_input["host"],
+                        CONF_USERNAME: user_input["username"],
+                        CONF_PASSWORD: user_input["password"],
+                        CONF_CONTROLLER_UNIQUE_ID: controller_unique_id,
+                    },
+                )
```
```suggestion
            if not errors:            
                controller_unique_id = hub.return_controller_unique_id()
                control4, model, mac = controller_unique_id.split("_", 3)
                formatted_mac = format_mac(mac)
                await self.async_set_unique_id(formatted_mac)
                self._abort_if_unique_id_configured()
                return self.async_create_entry(
                    title=controller_unique_id,
                    data={
                        CONF_HOST: user_input["host"],
                        CONF_USERNAME: user_input["username"],
                        CONF_PASSWORD: user_input["password"],
                        CONF_CONTROLLER_UNIQUE_ID: controller_unique_id,
                    },
                )
```

lawtancool committed:
add exception message to connect_to_director

bdraco commented:
This is getting close to mergeable 👍 

Please add `director_utils.py` to `.coveragerc`

```
pytest --cov=homeassistant/components/control4/ --cov-report term-missing -- tests/components/control4/test_*.py

Name                                                  Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------
homeassistant/components/control4/config_flow.py         85      9    89%   118-120, 147, 155, 159-185

```

Also there is still a bit of coverage missing for `config_flow.py`

lawtancool committed:
move C4Light creation outside of init

lawtancool committed:
fix config_flow for tests

Co-authored-by: J. Nick Koston <nick@koston.org>

lawtancool committed:
add director_utils to .coveragerc

lawtancool commented:
> pytest --cov=homeassistant/components/control4/ --cov-report term-missing -- tests/components/control4/test_*.py

Is it strictly necessary to increase the coverage? I think I could cover lines `158-184` by writing a test for the options flow, but other than that I don't see what else would really make sense.

Please correct me if I'm wrong, this is my first time working with Python tests! @bdraco

bdraco commented:
> > pytest --cov=homeassistant/components/control4/ --cov-report term-missing -- tests/components/control4/test_*.py
> 
> 
> 
> Is it strictly necessary to increase the coverage? I suppose I could cover lines `158-184` by writing a test for the options flow, but other than that I don't see what else would really make sense.
> 
> 
> 
> Please correct me if I'm wrong, this is my first time working with Python tests! @bdraco 

The goal is to have 100% coverage on config_flow.py,  I think the CI will still pass it at 96%

bdraco commented:
For `118-120` you can copy the invalid auth test and throw an exception like `ValueError` that won't be trapped to verify the broad exception handler is working in `async_step_user`

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,229 @@
+        controller_href
+    )
+
+    control4, model, mac_address = controller_unique_id.split("_", 3)
```
```suggestion
    _, model, mac_address = controller_unique_id.split("_", 3)
```

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,192 @@
+
+            if not errors:            
+                controller_unique_id = hub.return_controller_unique_id()
+                _, model, mac = controller_unique_id.split("_", 3)
```
```suggestion
                mac = (controller_unique_id.split("_", 3))[2]
```

bdraco commented on homeassistant/components/control4/light.py:
```diff
@@ -0,0 +1,225 @@
+        """Flag supported features."""
+        flags = 0
+        if self._is_dimmer:
+            flags |= SUPPORT_BRIGHTNESS
+            flags |= SUPPORT_TRANSITION
```
```suggestion
            flags |= SUPPORT_BRIGHTNESS | SUPPORT_TRANSITION
```

lawtancool committed:
Update homeassistant/components/control4/light.py

Co-authored-by: J. Nick Koston <nick@koston.org>

lawtancool committed:
remove unused variables from split

Co-authored-by: J. Nick Koston <nick@koston.org>

lawtancool committed:
remove unused variable from split

Co-authored-by: J. Nick Koston <nick@koston.org>

lawtancool committed:
add option_flow test

lawtancool committed:
Merge branch 'dev' of https://github.com/lawtancool/core into dev

lawtancool commented:
`config_flow.py` now has 100% test coverage! @bdraco 

```
pytest --cov=homeassistant/components/control4/ --cov-report term-missing -- tests/components/control4/test_*.py

Name                                               Stmts   Miss  Cover   Missing
--------------------------------------------------------------------------------
homeassistant/components/control4/config_flow.py      85      0   100%
```

lawtancool committed:
await super().async_added_to_hass()

bdraco commented:
> `config_flow.py` now has 100% test coverage! @bdraco
> 
> ```
> pytest --cov=homeassistant/components/control4/ --cov-report term-missing -- tests/components/control4/test_*.py
> 
> Name                                               Stmts   Miss  Cover   Missing
> --------------------------------------------------------------------------------
> homeassistant/components/control4/config_flow.py      85      0   100%
> ```

Nice.  I'll look before I get to bed which is soon (HST).

bdraco commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,229 @@
+    )
+    hass.data[DOMAIN][entry.entry_id][CONF_CONFIG_LISTENER]()
+    if unload_ok:
+        entry_id = entry.entry_id
+        hass.data[DOMAIN].pop(entry.entry_id)
+        _LOGGER.debug("Unloaded entry for %s", entry_id)
```
```suggestion
        hass.data[DOMAIN].pop(entry.entry_id)
        _LOGGER.debug("Unloaded entry for %s", entry.entry_id)
```

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,192 @@
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.device_registry import format_mac
+
+from .const import (
+    CONF_CONTROLLER_UNIQUE_ID,
+    CONF_LIGHT_COLD_START_TRANSITION_TIME,
+    CONF_LIGHT_TRANSITION_TIME,
+    DEFAULT_LIGHT_COLD_START_TRANSITION_TIME,
+    DEFAULT_LIGHT_TRANSITION_TIME,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    MIN_SCAN_INTERVAL,
+)
```
```suggestion
from .const import (
    CONF_CONTROLLER_UNIQUE_ID,
    CONF_LIGHT_COLD_START_TRANSITION_TIME,
    CONF_LIGHT_TRANSITION_TIME,
    DEFAULT_LIGHT_COLD_START_TRANSITION_TIME,
    DEFAULT_LIGHT_TRANSITION_TIME,
    DEFAULT_SCAN_INTERVAL,
    MIN_SCAN_INTERVAL,
)
from .const import DOMAIN  # pylint:disable=unused-import
```

lawtancool replied:
Good catch - I was wondering why pylint was complaining about an import I was actually using 😅

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,192 @@
+        except (Unauthorized, ClientError, asyncioTimeoutError):
+            _LOGGER.error("Failed to connect to the Control4 controller")
+            return False
+
+    def return_controller_unique_id(self) -> str:
+        """
+        Return the controller name found by authenticate().
+
+        This exists so that the controller name return value can be mocked in tests.
+        """
+        return self.controller_unique_id
```
```suggestion
```

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,192 @@
+                errors["base"] = "unknown"
+
+            if not errors:
+                controller_unique_id = hub.return_controller_unique_id()
```
```suggestion
                controller_unique_id = self. controller_unique_id
```

bdraco commented:
Should be good to merge once the CI is passing reliably

bdraco commented:
If token expire time is being checked you could do something like
```
diff --git a/tests/components/control4/test_config_flow.py b/tests/components/control4/test_config_flow.py
index a310baa4aa..5c51e96bc6 100644
--- a/tests/components/control4/test_config_flow.py
+++ b/tests/components/control4/test_config_flow.py
@@ -2,6 +2,7 @@
 import datetime
 
 from pyControl4.error_handling import Unauthorized
+import homeassistant.util.dt as dt_util
 
 from homeassistant import config_entries, setup
 from homeassistant.components.control4.const import (
@@ -31,7 +32,7 @@ def _get_mock_c4_account(
     },
     getDirectorBearerToken={
         "token": "token",
-        "token_expiration": datetime.datetime(2020, 7, 15, 13, 50, 15, 26940),
+        "token_expiration": dt_util.utcnow() + datetime.timedelta(minutes=60),
     },
 ):
     c4_account_mock = AsyncMock()
```

bdraco commented:
Misfired on the close and comment instead of comment button.  Sorry about that.

balloob commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,229 @@
+    account = C4Account(config[CONF_USERNAME], config[CONF_PASSWORD])
+    try:
+        await account.getAccountBearerToken()
+    except client_exceptions.ClientError as exception:
```
Would it be possible to distinguish about expired authentication and unable to reach the server?

lawtancool replied:
This step is retrieving an authentication token using the stored username and password, so I'll try to implement a check for invalid username/password.

balloob replied:
Would it be possible to not have to store user/pass and just keep tokens around ?

lawtancool replied:
I don't think so - even though the API says that account tokens never expire, I've found that sometimes they stop working, requiring re-authentication. 

The Control4 API is not publicly documented, so its safer to allow the integration to store the username/password and re-auth every time.

balloob commented on homeassistant/components/control4/light.py:
```diff
@@ -0,0 +1,224 @@
+            if ATTR_TRANSITION in kwargs:
+                transition_length = kwargs[ATTR_TRANSITION] * 1000
+            else:
+                transition_length = self._cold_start_transition_time * 1000
```
We shouldn't implement this at a platform level. This kind of feature should be added to the light integration so it can apply to any integration. It also doesn't make much sense that something called "start" is used for turn off.

lawtancool replied:
I agree - it seems that a default transition function could be implemented in the light profiles system https://github.com/home-assistant/core/blob/master/homeassistant/components/light/__init__.py#L143

balloob replied:
I even think that we have a PR open for that #36747

lawtancool replied:
Hey that's nice! I'll remove the default transition functionality + option flow from this PR then. In the future the light integration could be updated to support different "cold start/already on" transition times.

balloob commented on homeassistant/components/control4/light.py:
```diff
@@ -0,0 +1,224 @@
+            if ATTR_TRANSITION in kwargs:
+                transition_length = kwargs[ATTR_TRANSITION] * 1000
+            elif self.brightness == 0:
+                transition_length = self._cold_start_transition_time * 1000
```
I suggest we remove this option from this PR and explore how to add this feature to the light integration.

balloob commented on homeassistant/components/control4/strings.json:
```diff
@@ -0,0 +1,34 @@
+    "step": {
+      "user": {
+        "data": {
+          "host": "Controller hostname",
```
Interesting that we need to ask this. Can't we get the host from the cloud or via a scan?

lawtancool replied:
The Control4 controller does expose itself via SSDP, but it doesn't provide its own IP address in the SSDP `LOCATION` information. 

Is there a way to find the IP address a SSDP `NOTIFY` packet originates from without using the `LOCATION` info?

balloob replied:
Weird they don't have a location. I  think that we should be able to get the IP, but not sure if the current SSDP implementation offers that. It's fine like this for now, we'll ponder about it later :)

balloob commented on homeassistant/components/control4/strings.json:
```diff
@@ -0,0 +1,34 @@
+      "user": {
+        "data": {
+          "host": "Controller hostname",
+          "password": "Password",
```
This should be a reference.

Probably most other strings can also be using references. Here are some references that are part of the standard template.

```python
{
                "step": {
                    "user": {
                        "data": {
                            "host": "[%key:common::config_flow::data::host%]",
                            "username": "[%key:common::config_flow::data::username%]",
                            "password": "[%key:common::config_flow::data::password%]",
                        },
                    }
                },
                "error": {
                    "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
                    "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
                    "unknown": "[%key:common::config_flow::error::unknown%]",
                },
                "abort": {
                    "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
                },
            }
```

lawtancool replied:
I tried using references and then doing `python3 -m script.translations develop` to compile the strings, but the references aren't actually replaced by the strings they're supposed to be.

For example, I'll end up seeing "[%key:common::config_flow::data::host%]" in the Home Assistant UI.

MartinHjelmare replied:
The references will work in the production build. We haven't fixed that in the development script yet.

balloob commented on homeassistant/components/control4/strings.json:
```diff
@@ -0,0 +1,34 @@
+          "password": "Password",
+          "username": "Username/Email"
+        },
+        "title": "Control4",
```
This is not necessary. If you omit it, it's automatically using the integration name.
```suggestion
```

balloob commented on homeassistant/components/control4/strings.json:
```diff
@@ -0,0 +1,34 @@
+        "data": {
+          "light_transition_time": "Default light brightness transition time (seconds)",
+          "light_cold_start_transition_time": "Default light cold start time (seconds)",
+          "scan_interval": "Seconds between updates"
```
Why would we make this configurable ? We should pick the most aggressive option that the hub can handle. We shouldn't leave it up to the user to pick (and mess up)

lawtancool replied:
I put this in because Control4 systems can vary in size greatly (from tens to hundreds to even thousands?) of devices, so a polling rate that works for most may lag a older controller with many devices attached to it.

balloob replied:
Could we make some logic where we say

```
if devices < 50:
  return 5 seconds
elif devices < 200:
  return 15 seconds
else:
  return 1 minute
```

?

lawtancool replied:
I would rather not, mainly because I'm not sure how big a Control4 system would have to be before slowing down. It probably also depends on what specific features are being used (for example, surveillance video recording or movie playback).

I've changed the default to be more aggressive (5 seconds),  which I think should be appropriate for most systems. Most users should never have to change the scan_interval option now.

lawtancool committed:
disable pylint on DOMAIN import

Co-authored-by: J. Nick Koston <nick@koston.org>

lawtancool committed:
update log message when unloaded entry

Co-authored-by: J. Nick Koston <nick@koston.org>

lawtancool committed:
remove return_controller_unique_id()

not used for tests anymore

Co-authored-by: J. Nick Koston <nick@koston.org>

lawtancool committed:
remove unnecessary title from config_flow form

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

lawtancool committed:
remove light default transition

lawtancool committed:
change default scan interval to 5

lawtancool committed:
add check for incorrect user credentials

lawtancool commented:
I'm getting a `TypeError: object MagicMock can't be used in 'await' expression` error from pytest in the GitHub CI runs, but I've removed usage of MagicMock already and the test passes on my local system. The CI seems to be using an older version of `test_config_flow.py`.

balloob commented:
Make sure you import `patch` etc from `tests.async_mock`. There is a difference between py37 and 38.

lawtancool commented:
> Make sure you import `patch` etc from `tests.async_mock`. There is a difference between py37 and 38.

Yep, I am doing that. Is there some other difference between py37 and py38 I should be looking out for?
```
from tests.async_mock import AsyncMock, patch
from tests.common import MockConfigEntry
```

balloob commented:
Yes there is, `AsyncMock` behaves slightly different in edge cases. You can test locally using `pyenv`, which supports multiple Python versions. 

The difference is that on 3.7 we use `asynctest` and on 3.8 we use built-in `unittest.mock` which was upgraded to async.

balloob commented on tests/components/control4/test_config_flow.py:
```diff
@@ -0,0 +1,199 @@
+    },
+):
+    c4_account_mock = AsyncMock()
+    type(c4_account_mock).getAccountControllers = AsyncMock(
```
I think that this is wrong. 

```suggestion
    c4_account_mock.getAccountControllers.return_value = getAccountControllers
```

Does this work ?

balloob replied:
And since you want to mock a constructor, you can patch the return value. Updated above suggestion.

lawtancool replied:
```c4_account_mock.getAccountControllers.return_value = getAccountControllers``` ended up working. I'm going to push this change and see if that fixes py37 CI test.

lawtancool replied:
I'm still getting `TypeError: object MagicMock can't be used in 'await' expression` on py37, even though I've never imported or used `MagicMock` anywhere. 

I see that `CoroutineMock` should be imported for Python < 3.8
```
if sys.version_info[:2] < (3, 8):
    from asynctest.mock import *  # noqa
    from asynctest.mock import CoroutineMock as AsyncMock  # noqa
else:
    from unittest.mock import *  # noqa
```
and if that were true then there should be no way for `MagicMock `to appear since `CoroutineMock `inherits directly from `Mock `in the `asynctest `package.
```
class CoroutineMock(Mock):
```
This leads me to think that `AsyncMock` is incorrectly being imported directly from `unittest.mock` even on Python 3.7.8, as that inherits from some "magic" things:
```
class AsyncMock(AsyncMockMixin, AsyncMagicMixin, Mock):
```
What do you think?

lawtancool replied:
Never mind, I get the same error even when explicitly importing `from asynctest.mock import CoroutineMock as AsyncMock`

lawtancool committed:
fix test for py37

lawtancool committed:
add different test logic for py37 and py38

lawtancool committed:
isort test_config_flow.py

lawtancool commented:
@balloob I resolved the py37 test issue by adding separate test logic for py37 and py38. https://github.com/home-assistant/core/pull/37632/files#diff-062f72e4aaf9d728ab47d7e8abc47a91R34

Is this acceptable? I couldn't figure out how else to do it.

MartinHjelmare commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,225 @@
+    director_all_items = hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR_ALL_ITEMS]
+    return_list = []
+    for item in director_all_items:
+        if "categories" in item.keys() and category in item["categories"]:
```
We don't need to specify keys for membership test of dicts. That's the default. We can remove `.keys()`.

MartinHjelmare commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,225 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
```
We can remove the `CONFIG_SCHEMA`. It's not used.

MartinHjelmare commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,225 @@
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant. Pass in the `entry_data` instead.

MartinHjelmare commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,163 @@
+        self.account = None
+        self.controller_unique_id = None
+        self.director_bearer_token = None
+        self.director = None
```
The `account` and `director` variables don't need to be instance attributes as they are local to the respective methods.

MartinHjelmare commented on homeassistant/components/control4/light.py:
```diff
@@ -0,0 +1,206 @@
+    def brightness(self):
+        """Return the brightness of this light between 0..255."""
+        if self._is_dimmer:
+            return self._coordinator.data[self._idx]["value"] * 2.55
```
Do we need to round this value to an integer?

lawtancool replied:
I'm not sure what you mean - this seems to work fine in my local HA instance as is? Should I be copying this to `int`?

MartinHjelmare replied:
The brightness should be an integer between 0-255. Currently we're returning a float. I suggest rounding the value to an integer with `round` and no decimals.
```suggestion
            return round(self._coordinator.data[self._idx]["value"] * 2.55)
```

MartinHjelmare commented on homeassistant/components/control4/light.py:
```diff
@@ -0,0 +1,206 @@
+    @property
+    def is_on(self):
+        """Return whether this light is on or off."""
+        return bool(self._coordinator.data[self._idx]["value"] > 0)
```
We don't need to copy to bool. The expression is already a boolean.

MartinHjelmare commented on homeassistant/components/control4/manifest.json:
```diff
@@ -0,0 +1,16 @@
+      "st": "c4:director"
+    }
+  ],
+  "zeroconf": [],
```
Please remove empty items.

MartinHjelmare commented on tests/components/control4/test_config_flow.py:
```diff
@@ -0,0 +1,211 @@
+):
+    c4_account_mock = AsyncMock()
+
+    if sys.version_info[:2] < (3, 8):
```
This shouldn't be needed. Try speccing the mock with the class. Use the `spec=` keyword. That should set the correct awaitable mock methods.

https://docs.python.org/3/library/unittest.mock.html#unittest.mock.AsyncMock

lawtancool committed:
set spec of AsyncMock to avoid py37 issues

lawtancool committed:
remove empty items from manifest

lawtancool committed:
remove unnecessary copy to bool

lawtancool committed:
remove unnecessary instance variables

lawtancool committed:
pass in entry_data instead of hass

lawtancool committed:
remove CONFIG_SCHEMA

lawtancool committed:
remove unnecessary .keys()

lawtancool committed:
use string references

lawtancool committed:
round light brightness to integer

MartinHjelmare commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,221 @@
+    entry_data = hass.data[DOMAIN].setdefault(entry.entry_id, {})
+
+    config = entry.data
+    account = C4Account(config[CONF_USERNAME], config[CONF_PASSWORD])
```
Since the library is using aiohttp, we recommend it to accept an optional aiohttp client session argument. We want all integrations that use aiohttp to use the home assistant aiohttp client session. It can be retrieved using `async_get_clientsession` in the `aiohttp_client` helper module.
https://github.com/home-assistant/core/blob/8beaccf2ddb5a6e67619738996b86420b7ba831e/homeassistant/helpers/aiohttp_client.py#L32-L36

lawtancool replied:
Is it necessary to implement this in `config_flow.py` as well? The `HomeAssistant` object doesn't seem to be passed into any `config_entries.ConfigFlow` functions, so I'm not sure how I would get the client session when validating config.

bdraco replied:
@philhawthorne `self.hass` should be set in the config flow

balloob replied:
`self.hass`

balloob replied:
oops my comment was on an open tab that was no longer updated with latest comments 😬

lawtancool replied:
Thanks for the advice everyone 😅 I've added this to `config_flow.py` now.

lawtancool committed:
use ClientSession from hass

lawtancool committed:
bump pyControl4 requirement to 0.0.6

balloob commented on homeassistant/components/control4/__init__.py:
```diff
@@ -0,0 +1,226 @@
+            "Error authenticating with Control4 account API, incorrect username or password: %s",
+            exception,
+        )
+        raise ConfigEntryNotReady
```
This should not be a `ConfigEntryNotReady` because we can't recover from this. For this PR, let's just `return False`. In a future PR you can start a new config flow to gather the necessary info to fix this config entry.

balloob replied:
```suggestion
        return False
```

lawtancool committed:
use hass ClientSession in config_flow

lawtancool committed:
fix username and password strings

bdraco commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -36,18 +37,20 @@
     """Validates that config details can be used to authenticate and communicate with Control4."""
 
-    def __init__(self, host, username, password):
+    def __init__(self, host, username, password, hass: HomeAssistant):
```
If you are going to type, please type all the vars

lawtancool committed:
fail setup if Control4 credentials are incorrect

lawtancool committed:
remove type from Control4Validator init

balloob commented on homeassistant/components/control4/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+                await account.getDirectorBearerToken(self.controller_unique_id)
+            )["token"]
+            return True
+        except (Unauthorized, NotFound):
+            return False
```
It would be great if we could distinguish for the user in the error message if he has provided the incorrect credentials or if something could not be found.

lawtancool replied:
I think I'll just stop catching `NotFound` - I can't imagine a scenario where we just retrieved the controller name from Control4, then immediately after it can't be found.

balloob replied:
Just let the user know and they can handle it accordingly.

balloob commented on homeassistant/components/control4/director_utils.py:
```diff
@@ -0,0 +1,65 @@
+        await refresh_tokens(hass, entry)
+        director = hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR]
+        data = await director.getAllItemVariableValue(var)
+    return_dict = {}
+    for key in data:
+        return_dict[key["id"]] = key
+    return return_dict
```
Consider a dict comprehension
```suggestion
    return {key["id"]: key for key in data}
```

lawtancool committed:
use dict comprehension in director_update_data()

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

MartinHjelmare reviewed (approved):
Besides the scan interval option discussion this looks good.

balloob commented:
I'm fine with the scan interval being a choice. This looks great. Amazing first contribution @lawtancool 🐬

twotimesthree commented:
Wonderful job @lawtancool   Lights are working great!  Are contact sensors up next? I'd love to be able to see when doors are open/closed to tie them into my HA alarm.

I'm a web developer, is there anything I can do to assist?

twotimesthree commented:
@lawtancool Are you still the code owner for everything Control4?

frenck commented:
@lawtancool This is closed and handle pull request. If you want to report an bug, please create an issue. If you want to collaborate or need support, please use the community forums or Discord chat instead.

Thanks 👍 

../Frenck
